src/Groups/GAPGroups.jl

```diff
@@ -0,0 +1,617 @@
+# further possible functions: similar, literal_pow, parent_type
+
+export order, perm, cperm, isfinite, gens, ngens, comm, comm!, inv!, rand_pseudo, one!, div_right,
+div_left, div_right!, div_left!, deg, mul, mul!, listperm, degree, elements, right_coset, coset_decomposition,
+right_cosets , right_transversal, conjugacy_class, conjugacy_classes, number_conjugacy_classes, isconjugate, conjugacy_classes_subgroups, conjugacy_classes_maximal_subgroups, normalizer, normaliser, core, pcore, fitting_subgroup, frattini_subgroup, radical_subgroup, socle, sylow_subgroup, hall_subgroup, sylow_system, complement_system, hall_system, relators, isperfect, issimple, isalmostsimple, ispgroup
+
+
+# TODO: as soon as GAP packages like `polycyclic` or `rcwa` are loaded,
+# the custom group types and isos they define should be added to the arrays
+# _gap_group_types resp. _iso_function
+
+
+function group_element(G::T, x::GapObj) where T <: Group
+  return GAPGroupElem{T}(G, x)
+end
+
+function elements(G::T) where T <: Group
+  els = GAP.gap_to_julia(GAP.Globals.Elements(G.X))
+  elems = Vector{elem_type(G)}(undef, length(els))
+  i = 1
+  for x in els
+    elems[i] = group_element(G, x)
+    i += 1
+  end
+  return elems
+end
+
+# to be fixed later
+Base.hash(x::GAPGroupElem) = 0
+
+function parent(x::GAPGroupElem)
+  return x.parent
+end
+
+function Base.isfinite(G::PermGroup)
+  return true
+end
+
+function Base.isfinite(G::PcGroup)
+  return true
+end
+
+function Base.isfinite(G::Group)
+  return GAP.Globals.IsFinite(G.X)
+end
+
+function degree(x::PermGroup)
+   return x.deg
+end
+
+function order(x::Group)
+   return GAP.gap_to_julia(GAP.Globals.Size(x.X))
+end
+
+function order(x::GAPGroupElem)
+   return GAP.gap_to_julia(GAP.Globals.Order(x.X))
+end
+
+function order(::Type{T}, x::Union{GAPGroupElem, Group}) where T<:Number
+   return T(order(x))
+end
```

Comment:
The above `order` methods have it backwards: the base method should be the one which accept `::Type{T}`, and the one with a single argument should defer to the two-argument version with a default value for that type. That default value remains to be fixed (it could be BigInt, or Int64, or fmpz; might also have to be some new type which is a Union of one of the named types, plus a value for "infinity").

So something roughly like this (untested, needs work):
```julia
function order(::Type{T}, x::Union{GAPGroupElem, Group}) where T<:Number
    return GAP.gap_to_julia(T, GAP.Globals.Size(x.X))
end

function order(x::Union{GAPGroupElem, Group})
    return GAP.gap_to_julia(BigInt, GAP.Globals.Size(x.X)) # FIXME: use fmpz once we have a constructor / converter for that
end
```

Reply:
`function order`

Reply:
> function order

Reply:
The problem to fix a default value is that it returns an error instead of just the GAP object infinity. Should I define a new type containing all integers and a constant "infinity"?

Reply:
Well, that's one option. Another is to require people to always check `isfinite` first before calling `order`.

Personally, I am used to alway just calling `order` / `Size`; it also makes some code more elegant (e.g. checking if a group contains at least 5 elements is `Size(G) >= 5` in GAP, which is much more readable than `!isfinite(G) or order(G) >= 5`.

The agreement between OSCAR PIs so far was that we want to standardize on `fmpz` as default integer type in Oscar.jl. I guess that means we could add a new type, say `Order`, which is something like union of `fmpz` with a type providing `infinity` (we might also want to add in `-infinity` -- this is not needed for orders, but useful for e.g. degrees).

We could also (ab) use `Inf` from `Float64` for this, i.e. use `Union{fmpz, Float64}` as default type for the return value of `order`.

Or this could be built right into the `fmpz` type (I find it interesting that `Inf` and `-Inf` are apparently "real numbers" in Julia... ah well ;-).

Reply:
And if we simply return `Inf` when the group is infinite, even if it is a Float64 value, is there any problem in doing this? I mean, after checking all input types of functions that may use the group order...